tests/Credentials/ExternalAccountCredentialsTest.php

```diff
@@ -326,15 +318,198 @@ public function testFetchAuthTokenWithImpersonation()
+                    $responseBody = 'abc';
+                    break;
+                case 2:
+                    $this->assertEquals('token-url.com', (string) $request->getUri());
+                    $responseBody = '{"access_token": "def"}';
+                    break;
+                case 3:
+                    $this->assertEquals(
+                        'https://cloudresourcemanager.googleapis.com/v1/projects/' . $expectedProjectNumber,
+                        (string) $request->getUri()
+                    );
+                    $responseBody = json_encode(['projectId' => 'test-project-id']);
+                    break;
+            }
+
+            $body = $this->prophesize(StreamInterface::class);
+            $body->__toString()->willReturn($responseBody);
+
+            $response = $this->prophesize(ResponseInterface::class);
+            $response->getBody()->willReturn($body->reveal());
+            $response->hasHeader('Content-Type')->willReturn(false);
+
+            return $response->reveal();
+        };
+
+        $creds = new ExternalAccountCredentials('a-scope', $jsonCreds);
+        $this->assertEquals('test-project-id', $creds->getProjectId($httpHandler));
+    }
+
+    public function provideGetProjectId()
+    {
+        return [
+            // from audience
+            [
+                [
+                    'audience' => '//iam.googleapis.com/projects/1234/locations/global/workloadIdentityPools/foo/providers/bar',
+                ] + $this->baseCreds,
+                '1234'
+            ],
+            // from workforce_pool_user_project
+            [
+                [
+                    'audience' => '//iam.googleapis.com/locations/global/workforcePools/foo/providers/bar',
+                    'workforce_pool_user_project' => '4567',
+                ] + $this->baseCreds,
+                '4567'
+            ],
+        ];
+    }
+
+    public function testCacheIsCalledForGetProjectIdWithCache()
+    {
+        $jsonCreds = [
+            'audience' => '//iam.googleapis.com/projects/1234/locations/global/workloadIdentityPools/foo/providers/bar',
+        ] + $this->baseCreds;
+
+        $httpHandler = function (RequestInterface $request) {
+            $this->assertEquals(
+                'https://cloudresourcemanager.googleapis.com/v1/projects/1234',
+                (string) $request->getUri()
+            );
+            $this->assertEquals('Bearer some-token', $request->getHeaderLine('authorization'));
+            $body = $this->prophesize(StreamInterface::class);
+            $body->__toString()->willReturn(json_encode(['projectId' => 'test-project-id']));
+
+            $response = $this->prophesize(ResponseInterface::class);
+            $response->getBody()->willReturn($body->reveal());
+            $response->hasHeader('Content-Type')->willReturn(false);
+
+            return $response->reveal();
+        };
```

Comment:
Can you help me understand why are we defining this httphandler if the token and eventually the project id would be fetched from the mocked cache?

Reply:
Sure. This method is testing the getProjectId method, which makes an API call using the project number in order to retrieve the project ID. It's also testing that a cached access token is used when that API call is made, instead of retrieving a new one.

So the cache is for the access token, which is needed for the API request to get the project ID. The httpHandler is the HTTP request made when getProjectId is called, which is not cached.

Reply:
Oh, now this makes sense. 
I feel adding a small explanation there in comment would go a long way to quickly understand it.